### PR TITLE
refactor(conditional-page-build): tracking PR

### PR DESCRIPTION
**Pull request by pieh**
## Description

This is tracking branch for moving `GATSBY_EXPERIMENTAL_PAGE_BUILD_ON_DATA_CHANGES` from experimental to stable

Changes are being plucked from this massive PR into smaller ones:

Already merged:
 - https://github.com/gatsbyjs/gatsby/pull/29396 - `refactor(html-ssr): remove fs from ssr bundle, remove static queries from ssr bundle`
 -  https://github.com/gatsbyjs/gatsby/pull/29473 - `refactor(conditional-page-build): explicitly persist status of generated html files` (this just refactor current `GATSBY_EXPERIMENTAL_PAGE_BUILD_ON_DATA_CHANGES` to structure that will allow us to track additional things and remove some problems of having to constatly use `GATSBY_EXPERIMENTAL_PAGE_BUILD_ON_DATA_CHANGES` and not switch between build and develop)
 - https://github.com/gatsbyjs/gatsby/pull/29482 - `feat(conditional-page-builds): track ssr compilation hash`
 - https://github.com/gatsbyjs/gatsby/pull/29486 - `feat(conditional-page-builds): track static query results`
 - https://github.com/gatsbyjs/gatsby/pull/29498 - `fix(conditional-page-builds): track page data and not page query`
 - https://github.com/gatsbyjs/gatsby/pull/29499 - `chore(build): move html generation to separate function`
 
Ready to review (they are serial - each next one builds on previous and target previous PR branch, so those should be merged in order):
- https://github.com/gatsbyjs/gatsby/pull/29548 - `feat(conditional-page-builds): make it default`

Future PRs:
 - feature
   - wrap `fs`, `http`, `child_process` node builtin modules in SSR bundle and in case of use of those in ssr generating, mark it's not safe to use conditional page builds and fallback to building every page ( :( ). This is somewhat waiting for webpack@5 bump, as it's not straight forward to do in webpack@4 and there are changes in webpack@5 that any efforts made for webpack@4 might be rendered void
 - fixes:
   - clear `require.cache` from ssr renderer if it changed
 - cleanups:
   - remove "dead" `setPageData` action and types
   - https://github.com/gatsbyjs/gatsby/pull/29498#discussion_r576984765
   - maybe https://github.com/gatsbyjs/gatsby/pull/29473#discussion_r576342167 - this require more effort than just cleanup
   - https://github.com/gatsbyjs/gatsby/pull/29548#discussion_r577589506

### CODE REVIEW

File: docs/docs/conditional-page-builds.md [hunk 1]
```diff
@@ -50,12 +50,8 @@ Done in 154.501 sec
...
 
-- This feature is not available with `gatsby develop`.
-
 - You should not try to use this flag alongside Incremental Builds in Gatsby Cloud, as it uses a different process and may conflict with it.
 
 - You will need to persist the `.cache` and `public` directories between builds. This allows for comparisons and reuse of previously built files. If `.cache` directory was not persisted then a full build will be triggered. If `public` directory was not persisted then you might experience failing builds or builds that are missing certain assets.
 
 - Any code changes (templates, components, source handling, new plugins etc) will prompt the creation of a new webpack compilation hash and trigger a full build.
-
-Note: When using the `GATSBY_EXPERIMENTAL_PAGE_BUILD_ON_DATA_CHANGES` flag it is important to do so consistently when building your project. Otherwise, the cache will be cleared and the necessary data for comparison will no longer be available, removing the ability to check for incremental data changes.
```
Comment (pieh) [inline docs/docs/conditional-page-builds.md:61]: those changes will need to be extracted to separate PR and only applied when stable version including code changes is released

File: packages/gatsby/src/utils/page-html.ts [hunk 2]
```diff
@@ -28,3 +30,25 @@ export const remove = async (
...
+export function calcDirtyHtmlFiles(
+  state: IGatsbyState
+): { toRegenerate: Array<string>; toDelete: Array<string> } {
+  const toRegenerate: Array<string> = []
+  const toDelete: Array<string> = []
+
+  state.html.trackedHtmlFiles.forEach(function (htmlFile, path) {
+    if (htmlFile.isDeleted || !state.pages.has(path)) {
+      // FIXME: checking pages state here because pages are not persisted
+      // and because of that `isDeleted` might not be set ...
```
Comment (pieh) [inline packages/gatsby/src/utils/page-html.ts:43]: Related:
 - https://github.com/gatsbyjs/gatsby/pull/28590
 - https://github.com/gatsbyjs/gatsby/pull/28760

**Review by KyleAMathews — COMMENTED** [review]
Looking 👍 

left a few comments

File: packages/gatsby/cache-dir/static-entry.js [hunk 3]
```diff
@@ -77,6 +80,20 @@ const getPageData = pagePath => {
   }
 }
 
+// TODO: verify if memoization is safe given that results can change
+const getStaticQueryResult = memoize(hash => {
```
[the remark this answers is not in the document]
Comment (pieh) [reply packages/gatsby/cache-dir/static-entry.js:84]: Yeah, that's the actual thing to verify ;) depends if ssr bundle is "reloaded" or just kept in memory

File: packages/gatsby/src/commands/build-html.ts [hunk 4]
```diff
@@ -152,6 +162,12 @@ const renderHTMLQueue = async (
       paths: pageSegment,
     })
 
+    // is this code path used for dev ssr?
+    store.dispatch({
```
[the remark this answers is not in the document]
Comment (pieh) [reply packages/gatsby/src/commands/build-html.ts:166]: That's good, but I will add test case for it with DEV_SSR enabled to make sure of that (+ shield from potential regressions in future)

**Comment by pieh** [comment]
This tracking PR served it's purpose and all that actually ment to be tracked with this PR was merged, so it's to close this one
